### ISSUE TITLE
Rust Cross compile to embedded target fails due to non working rustc

### ISSUE THREAD

**Issue by bauen1** [mesonbuild/meson#9111]
**Describe the bug**
Trying to cross compile with the rust language to an embedded target (no std support), fails at the setup step.

I think adding `#![no_std]` to https://github.com/mesonbuild/meson/blob/a9e9b7c7501a3c8a5984a93879d1f309bf8c72aa/mesonbuild/compilers/rust.py#L69-L72 would solve the issue.

**To Reproduce**

meson.build:
```
project('bare-metal', ['rust'])

static_library('baremetal', 'src/lib.rs', rust_crate_type: 'staticlib')
```

cross.ini
```
[binaries]
c = '/usr/bin/arm-none-eabi-gcc'
rust_ld = '/usr/bin/arm-none-eabi-gcc'
rust = ['rustc', '--target', 'thumbv7m-none-eabi']

[host_machine]
system = 'lm3s6965evb'
cpu_family = 'arm'
cpu = 'thumbv7'
endian = 'little'
```

**Expected behavior**
It compiles.

**Comment by PThierry** [comment]
Hello ! Same problem here. Though the no-std is a cross-compilation problem, not native one. I would better say something like:
```python
        with open(source_name, 'w', encoding='utf-8') as ofile: 
            if self.is_cross:
                ofile.write(textwrap.dedent(
                    '''#![no_main]
                    #![no_std]
                    use core::panic::PanicInfo;
                    #[panic_handler]
                    fn panic(_panic: &PanicInfo<'_>) -> ! {
                        loop {}
                    }
                    '''))
            else:
                ofile.write(textwrap.dedent(
                    '''fn main() { 
                    }
                    '''))
```

**Comment by xclaesse** [comment]
I agree `#![no_std]` makes sense for a sanity check, we just want to make sure the compiler is somewhat functional, like detect if it does not exist at all, or user messed up and wrote `rust=gcc` in cross file, etc. I think that's fine even for native compilation, a project that knows they are not going to use the std could not install it even for native platform, e.g. to reduce docker image size on CI, right?

However a `main()` function is still good to have, because you could be cross compiling and have an exe_wrapper (e.g. cross build for Windows with wine). In that case we want to ensure that very basic executable can be run. Unless there is a reason why having a `main()` would not work in legitimate use-case?

Mind making a PR?

**Comment by PThierry** [comment]
I'm testing by know, but havin main (see below) in no-std mode seems problematic as it requires lang_items feature... and this require a +nighly selector... :-/
Here is the current snippet :
```python
        with open(source_name, 'w', encoding='utf-8') as ofile: 
            if self.is_cross:
                ofile.write(textwrap.dedent(
                    '''#![no_std]
                    #![feature(lang_items)]
                    use core::panic::PanicInfo;

                    #[lang = "eh_personality"]
                    extern "C" fn eh_personality() {}
                    fn main() {
                    }
                    #[panic_handler]
                    fn panic(_panic: &PanicInfo<'_>) -> ! {
                        loop {}
                    }
                    '''))
            else:
                ofile.write(textwrap.dedent(
                    '''fn main() { 
                    }
                    '''))

        cmdlist = self.exelist + ['-o', output_name, source_name]
```
This requires the cmdlist to be updated in consequence

**Comment by xclaesse** [comment]
Note that we don't currently have Rust cross compilation CI. I tried to add one and it exploded in a million places: https://github.com/mesonbuild/meson/pull/11990.

**Comment by PThierry** [comment]
> Note that we don't currently have Rust cross compilation CI. I tried to add one and it exploded in a million places: #11990.

Ack :-) I will try locally in my own Docker and take you informed :-)

**Comment by JamBalaya56562** [comment]
Meson's Rust sanity check now detects bare-metal targets and no longer forces a `fn main()` program that requires std.

Verification: See commit 1aac6cc1ec380a14519822d1a2bdf3e5850de68b ("compiler,rust: fix sanity check and internal deps for baremetal rust project/target"). In current master, `RustCompiler._sanity_check_source_code()` in `mesonbuild/compilers/rust.py` returns a `#![no_std] #![no_main]` program with a `#[panic_handler]` (and `pub fn _start()`) when the host `kernel == 'none'`, and `_sanity_check_compile_args()` adds `-nostartfiles` to the linker for that case. Cross-compiling a Rust `staticlib` to a target like `thumbv7m-none-eabi` should now pass the setup step, which is what this issue asked for. This can be closed.
